Svc/BufferManager/BufferManagerComponentImpl.cpp

```diff
@@ -0,0 +1,219 @@
+    NATIVE_UINT_TYPE memorySize = 0; // track needed memory
+    this->m_numStructs = 0; // size the number of tracking structs
+    // walk through bins and add up the sizes
+    for (NATIVE_UINT_TYPE bin = 0; bin < BUFFERMGR_MAX_NUM_BINS; bin++) {
+        if (this->m_bufferBins.bins[bin].numBuffers) {
+            memorySize += 
+                this->m_bufferBins.bins[bin].bufferSize * this->m_bufferBins.bins[bin].numBuffers // allocate each set of buffer memory
+                + sizeof(AllocatedBuffer) * this->m_bufferBins.bins[bin].numBuffers; // allocate the structs to track the buffers
+            this->m_numStructs += this->m_bufferBins.bins[bin].numBuffers;
+        }
+    }
+
+    NATIVE_UINT_TYPE allocatedSize = memorySize;
+    bool recoverable; //!< don't care if it is recoverable since they are a pool of user buffers
+
+    // allocate memory
+    void *memory = allocator.allocate(memId,allocatedSize,recoverable);
+    // make sure the memory returns was non-zero and the size requested
+    FW_ASSERT(memory);
+    FW_ASSERT(memorySize == allocatedSize,memorySize,allocatedSize);
+    // structs will be at beginning of memory
+    this->m_buffers = static_cast<AllocatedBuffer*>(memory);
+    // memory buffers will be at end of structs in memory, so compute that memory as the begining of the 
+    // struct past the number of structs
+    U8* bufferMem = reinterpret_cast<U8*>(&this->m_buffers[this->m_numStructs]);
+
+    // walk through entries and initialize them
+    NATIVE_UINT_TYPE currStruct = 0;
+    for (NATIVE_UINT_TYPE bin = 0; bin < BUFFERMGR_MAX_NUM_BINS; bin++) {
+        if (this->m_bufferBins.bins[bin].numBuffers) {
+            for (NATIVE_UINT_TYPE binEntry = 0; binEntry < this->m_bufferBins.bins[bin].numBuffers; binEntry++) {
+                // placement new for Fw::Buffer instance. We don't need the new() return value, 
+                // because we know where the Fw::Buffer instance is
+                U32 context = (this->m_mgrId << 16) | currStruct;
+                (void) new(&this->m_buffers[currStruct].buff) Fw::Buffer(bufferMem,this->m_bufferBins.bins[bin].bufferSize,context);
+                this->m_buffers[currStruct].allocated = false;
+                this->m_buffers[currStruct].memory = bufferMem;
+                this->m_buffers[currStruct].size = this->m_bufferBins.bins[bin].bufferSize;
+                bufferMem += this->m_bufferBins.bins[bin].bufferSize;
+                currStruct++;
+            }
+        }
+    }
+      
+    // check some assertions
+    FW_ASSERT(bufferMem == (static_cast<U8*>(memory) + memorySize));
+    FW_ASSERT(currStruct == this->m_numStructs,currStruct,this->m_numStructs);
+    // indicate setup is done
+    this->m_setup = true;
+  }
+
+  void BufferManagerComponentImpl ::
+    schedIn_handler(
+        const NATIVE_INT_TYPE portNum,
+        NATIVE_UINT_TYPE context
+    )
+  {
+    // write telemetry values
+    this->tlmWrite_HiBuffs(this->m_highWater);
```

Comment:
It may be a good idea to toggle these telemetry reports on/off via commands. By default, these telemetry channels could be turned off and not reported to avoid constantly generating telemetry packets. When needed, these channels can be turned on for reporting via a command. This would help control telemetry packet generation/output especially in radio RF bandwidth constrained environments.

Reply:
A related issue is that the previous version had a mechanism for suppressing out-of-memory warnings, so repeated requests wouldn't cause spam. I don't think that mechanism has been incorporated here. Maybe add event throttling to that event?

Reply:
Will be followed under issue https://github.com/nasa/fprime/issues/342